### DISCUSSION TITLE
Wave Function Collapse

### DISCUSSION THREAD

**Discussion by berinhard** [py5coding/py5generator#508] [Show and tell]
Hey people! I'm passing by as suggested by @villares to share with you my implementation of the Wave Function Collapse algorithm in Python on top of py5.  Of course, I was heavily influenced by the [coding traing challenge](https://thecodingtrain.com/challenges/171-wave-function-collapse) on this theme, but I went with some different data modeling decisions.

I feel that their solution is heavily based on loose objects and coupling between data structures indexes. For example, they represent a matrix of cells as a one-dimensional array. Although I know it makes sense in terms of performance and all of that, it's a strategy that hardly fits into my mind. I don't know, for me it's easier to read a code which says me something than doing math operations to understand the intentions of the code.

Guided by this principle, I came up with an more OOO-like implementation which, although is not the optimal in terms of performance, seemed kind of easier to grasp for me. My goal is to have a high-level API where I only need to define tiles with their adjacency rules and the dimension of the grid. In the end, the overall `setup`/`draw` methods looks like this:

```python
grid = None


def setup():
    global grid
    py5.size(800, 800)

    images_dir = Path.home() / "envs" / "sketches" / "be5" / "algorithms" / "tiles" / "demo"
    blank = Tile.from_file(images_dir / "0.png", edges=[0, 0, 0, 0])
    ptr_1 = Tile.from_file(images_dir / "1.png", edges=[0, 1, 1, 1])
    ptr_2 = Tile.from_file(images_dir / "2.png", edges=[0, 1, 0, 1])

    tiles = [
        blank,
        ptr_1,
        ptr_1.rotate(1),
        ptr_1.rotate(2),
        ptr_1.rotate(3),
        ptr_2,
        ptr_2.rotate(1),
    ]
    grid = WaveFunctionCollapseGrid(dim=40, tiles=tiles)
    grid.start()


def draw():
    grid.collapse()
    grid.draw()

    if grid.complete:
        print("finished!")
        py5.no_loop()


py5.run_sketch()
```

The final code be [found here](https://github.com/berinhard/sketches/blob/master/be5/algorithms/wave_function_collapse.py). Thie `be5` directory is a set of helpers that I've used to have with PyProcessing and which I'm porting to py5 now. My idea is to release it as an independent utilities belt to have alongside py5 so I can have some more complex tools, such as WFC algorithm, easily reachable.

There are still things that I wanna add to this code like, for example, to support non-symmetrical tiles rotations. But the base foundation for it is there. 

See you =)

**Comment by hx2A** [comment]
Hi, @berinhard ! Thank you so much for posting this. It means a lot to me when people share their py5-related joy!

> I don't know, for me it's easier to read a code which says me something than doing math operations to understand the intentions of the code.

Couldn't agree more. Sometimes it is best to write something that works and does the right thing first. Later, alter the code for performance. Then you have a slow but working version to compare the fast version with, and can confirm it is done correctly.

I don't know much about Wave Function Collapse but it looks interesting. What's the best way to run your code for this? It isn't clear from the code in that folder.

> Thie be5 directory is a set of helpers that I've used to have with PyProcessing and which I'm porting to py5 now. My idea is to release it as an independent utilities belt to have alongside py5 so I can have some more complex tools, such as WFC algorithm, easily reachable.

Great, that's excellent. BTW, py5 does have support for library extensions, just like Processing does. You can write a library that uses hooks to attach itself to a Sketch, telling py5 to call the library's functions before or after any user function such as `draw()`.  Right now this hook functionality is only used internally, and there are no library extensions built yet. But if you have an idea for one, let me know, I'll explain more about how it works.
